File: Design-Patterns-Using-Java/Momento/MomentoPatternJson.java
import java.util.ArrayList;
import java.util.List;

class Memento implements Cloneable {
    private Object state;

    public Memento(Object state) {
        this.state = state;
    }

    public Object getState() {
        return state;
    }

    @Override
    protected Object clone() throws CloneNotSupportedException {
        return super.clone();
    }
}

class Originator {
    private Object state;

    public void setState(Object state) {
        this.state = state;
    }

    public Object getState() {
        return state;
    }

    public Memento createMemento() {
        return new Memento(state);
    }

    public void setMemento(Memento m) {
        this.state = m.getState();
    }
}

class CareTaker {
    private List<Memento> history;
    private int top;
    private int max;

    public CareTaker() {
        this.history = new ArrayList<>();
        this.top = -1;
        this.max = -1;
    }

    public void addMemento(Memento m) {
        top += 1;
        max = top;
        if (top <= history.size() - 1) {
            try {
                history.set(top, (Memento) m.clone());
            } catch (CloneNotSupportedException e) {
                e.printStackTrace();
            }
        } else {
            try {
                history.add((Memento) m.clone());
            } catch (CloneNotSupportedException e) {
                e.printStackTrace();
            }
        }
    }

    public Memento getMemento(int index) {
        return history.get(index);
    }

    public Memento undo() {
        System.out.println("Undoing state.");
        if (top <= 0) {
            top = 0;
            return history.get(0);
        }

        top -= 1;
        return history.get(top);
    }

    public Memento redo() {
        System.out.println("Redoing state.");
        if (top >= (history.size() - 1) || top >= max) {
            return history.get(top);
        }

        top += 1;
        return history.get(top);
    }

    public int getStatesCount() {
        return history.size();
    }
}

public class MomentoPatternJson {
    public static void main(String[] args) {
        Originator originator = new Originator();
        CareTaker careTaker = new CareTaker();

        originator.setState("State 1");
        careTaker.addMemento(originator.createMemento());
        System.out.println(originator.getState());

         originator.setState("State 2");
        careTaker.addMemento(originator.createMemento());
        System.out.println(originator.getState());

        originator.setState("State 3");
        careTaker.addMemento(originator.createMemento());
        System.out.println(originator.getState());

        originator.setMemento(careTaker.undo());
        System.out.println(originator.getState());

        originator.setMemento(careTaker.undo());
        System.out.println(originator.getState());

        originator.setState("State 4");
        careTaker.addMemento(originator.createMemento());
        System.out.println(originator.getState());

        originator.setMemento(careTaker.redo());
        System.out.println(originator.getState());

        originator.setMemento(careTaker.redo());
        System.out.println(originator.getState());

        originator.setMemento(careTaker.redo());
        System.out.println(originator.getState());
    }
}
/* 
State 1
State 2
State 3
Undoing state.
State 2
Undoing state.
State 1
State 4
Redoing state.
State 4
Redoing state.
State 4
Redoing state.
State 4
*/

File: Design-Patterns-Using-Kotlin/Momento/MomentoPatternJson.kt
import java.util.ArrayList
import kotlin.collections.MutableList

data class Memento(val state: Map<String, Any?>)

class Originator {
    private var state: MutableMap<String, Any?> = mutableMapOf()

    fun setState(state: String) {
        this.state["state"] = state
    }

    fun getState(): Map<String, Any?> {
        return state
    }

    fun createMemento(): Memento {
        return Memento(HashMap(state))
    }

    fun setMemento(m: Memento) {
        println(m.state)
        state = HashMap(m.state)
    }
}

class CareTaker {
    private val history: MutableList<Memento> = ArrayList()
    private var top = -1
    private var max = -1

    fun addMemento(m: Memento) {
        top++
        max = top
        if (top <= history.size - 1) {
            history[top] = Memento(HashMap(m.state))
        } else {
            history.add(Memento(HashMap(m.state)))
        }
    }

    fun getMemento(index: Int): Memento {
        return history[index]
    }

    fun undo(): Memento {
        println("Undoing state.")
        if (top <= 0) {
            return Memento(HashMap(getMemento(0).state))
        }

        top--
        return Memento(HashMap(getMemento(top).state))
    }

    fun redo(): Memento {
        println("Redoing state.")
        if (top >= history.size - 1 || top >= max) {
            return Memento(HashMap(getMemento(top).state))
        }

        top++
        return Memento(HashMap(getMemento(top).state))
    }

    fun getStatesCount(): Int {
        return history.size
    }
}

// Client Code
fun main() {
    val originator = Originator()
    val careTaker = CareTaker()

    originator.setState("State 1")
    careTaker.addMemento(originator.createMemento())
    println(originator.getState())

    originator.setState("State 2")
    careTaker.addMemento(originator.createMemento())
    println(originator.getState())

    originator.setState("State 3")
    careTaker.addMemento(originator.createMemento())
    println(originator.getState())

    originator.setMemento(careTaker.undo())
    println(originator.getState())

    originator.setMemento(careTaker.undo())
    println(originator.getState())

    originator.setState("State 4")
    careTaker.addMemento(originator.createMemento())
    println(originator.getState())

    originator.setMemento(careTaker.redo())
    println(originator.getState())

    originator.setMemento(careTaker.redo())
    println(originator.getState())

    originator.setMemento(careTaker.redo())
    println(originator.getState())
}

/*
{state=State 1}
{state=State 2}
{state=State 3}
Undoing state.
{state=State 2}
{state=State 2}
Undoing state.
{state=State 1}
{state=State 1}
{state=State 4}
Redoing state.
{state=State 4}
{state=State 4}
Redoing state.
{state=State 4}
{state=State 4}
Redoing state.
{state=State 4}
{state=State 4}
*/